src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,295 @@
+
+            if (GetMethodName(syntaxFacts, expression) != nameof(string.Format))
+            {
+                return;
+            }
+
+            var arguments = syntaxFacts.GetArgumentsOfInvocationExpression(context.Node);
+            var numberOfArguments = arguments.Count;
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(expression, context.CancellationToken);
+
+            if (!TryGetValidFormatMethodSymbol(context, numberOfArguments, symbolInfo, out var method))
+            {
+                return;
+            }
+
+            var parameters = method.Parameters;
+
+            // If the chosen overload has a params object[] that is being passed an array of
+            // reference types then an actual array must be being passed instead of passing one
+            // argument at a time because we know the set of overloads and how overload resolution
+            // will work.  This array may have been created somewhere we can't analyze, so never
+            // squiggle in this case.
+            if (ArgsIsArrayOfReferenceTypes(context.SemanticModel, arguments, parameters))
+            {
+                return;
+            }
+
+            if (!TryGetFormatStringLiteralExpressionSyntax(
+                context.SemanticModel,
+                arguments,
+                parameters,
+                out var formatStringLiteralExpressionSyntax))
+            {
+                return;
+            }
+
+            if (parameters.Length == 0)
+            {
+                return;
+            }
+
+            var hasIFormatProvider = parameters[0].Type.Equals(formatProviderType);
+
+            // We know the format string parameter exists so numberOfArguments is at least one,
+            // and at least 2 if there is also an IFormatProvider.  The result can be zero if 
+            // calling string.Format("String only with no placeholders"), where there is an 
+            // empty params array.
+            var numberOfPlaceholderArguments = numberOfArguments - (hasIFormatProvider ? 2 : 1);
+
+            var formatString = GetLiteralExpressionSyntaxAsString(formatStringLiteralExpressionSyntax);
+
+            if (FormatCallWorksAtRuntime(formatString, numberOfPlaceholderArguments))
+            {
+                return;
+            }
+            
+            ValidateAndReportDiagnostic(context, numberOfPlaceholderArguments,
+                formatString, GetLiteralExpressionSyntaxSpanStart(formatStringLiteralExpressionSyntax));
+        }
+
+        private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode expression)
+        {
+            if (syntaxFacts.IsSimpleMemberAccessExpression(expression))
+            {
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).ValueText;
```

Comment:
Consider something like: `c.Method<int>("something")`.  Does `GetIdentifierOfSimpleName` work the way you expect int that case?